feat(snippet): Minimal snippet API

hrsh7th opened the pull request:
This PR implements the minimal snippet feature.

### API

- `vim.snippet.expand(snippet_text)`
- `vim.snippet.jump(direction)`
- `vim.snippet.jumpable(direction)`
- `vim.snippet.dispose()`

### Limitation
- current implementation can't handle adjacent placeholders.
- the `transform` feature is just ignored.

### TODO
- functionaltest
    - port test case from vim-vsnip
- documentation
    - learn how to write docs

hrsh7th commented on runtime/lua/vim/snippet.lua:
```diff
@@ -0,0 +1,200 @@
+  self.marks = traverse(snippet_node, cursor)
+end
+
+local M = {}
```
In my desing, the `vim.snippet` will be this `M`.

The plugin authors can override `vim.snippet.expand` or `vim.snippet.jump`.

Probably, we should expose `vim.snippet.jumpable(direction)`.

hrsh7th commented on runtime/lua/vim/snippet.lua:
```diff
@@ -0,0 +1,212 @@
+---@class nvim.SnippetSession
+---@field ns string
+---@field public snippet_text string
+local SnippetSession = {}
```
The default minimal snippet implementation is this.

hrsh7th commented on runtime/lua/vim/snippet.lua:
```diff
@@ -0,0 +1,212 @@
+  end
+end
+
+local M = {}
```
This `M` will be a `vim.snippet`.

The plugin authors can override `vim.snippet.expand` and `vim.snippet.jump`.

Probably, we should expose `vim.snippet.jumpable(direction)`.

hrsh7th commented on runtime/lua/vim/snippet.lua:
```diff
@@ -0,0 +1,212 @@
+
+---Jump to the suitable placeholder.
+---@param direction nvim.SnippetJumpDirection
+function SnippetSession:jump(direction)
```
This implementation does not care about the `current tabstop index`. Instead of that, this determines the suitable next tabstop via `cursor position`.

This is not general implementation. I think we should implement correct version if we ship this PR.

hrsh7th commented on runtime/lua/vim/snippet.lua:
```diff
@@ -0,0 +1,212 @@
+  cursor[1] = cursor[1] - 1
+
+  local snippet_node = snippet.parse(self.snippet_text)
+  vim.api.nvim_buf_set_text(0, cursor[1], cursor[2], cursor[1], cursor[2], vim.split(tostring(snippet_node), '\n'))
```
Currently, this will insert the result of `tostring(snippet_ast)`.
It should be changed when implementing LSP variables.

hrsh7th commented:
the `adjacent placeholder` is difficult to support via extmarks API, I guess.

I don't think the core needs to support such difficult cases. How about that?

hrsh7th commented:
Hm... the gravity management is bit hard. I'll investigate it...

hrsh7th commented:
Hmm ... undo complex extmarks don't restore state correctly, I guess

hrsh7th commented on test/functional/lua/snippet_spec.lua:
```diff
@@ -0,0 +1,106 @@
+    end
+  end)
+
+  -- This test doesn't work as expected due to feedkeys (maybe).
```
I can't manage this test...

hrsh7th commented:
The following are features that can be implemented but not implemented. How far should it be implemented?

1. Nested snippet expansion

This feature allows expanding the snippet even if the user has already in another snippet.
Typically, the two snippets will be merged as one snippet session and the user can jump to all merged placeholders.

2. Squash nested placeholders

This feature is only for nested placeholders. For example, the following snippets can be removed `extends Parent` with `<BS>`. In this case, another snippet engine (e.g. vim-vsnip) will remove the `Parent` placeholders.

```
class ${1:Class}${2: extends ${3:Parent}} {
  constructor($4) {
    $0
  }
}
```

clason commented:
I'd say the _minimal_ snippet API we should support is what is required for LSP completion snippets to work -- and no more. So if these features are not used by the LSP spec, we should not implement them.

Of course, a "nested" situation can also occur if an LSP snippet is expanded inside a user snippet -- but then the user must already be using a full snippet plugin, which should also take care of LSP snippets and so this is the full plugin's problem. (We should not support mixing different snippet engines!)

hrsh7th commented:
Thank you for your response.
OK. I'll make the implementation more simple.

clason commented:
Oh, I'm not saying you have to remove stuff!

hrsh7th commented:
Yes. I think the required features are

1. undo management
2. Support editing non-origin mark directly

Currently, my implementation cares about the adjacent placeholders. I'll remove it for now. (If we re-consider it's required, we can implement it later)

zeertzjq commented:
Rebasing on master may prevent test failure in `tui_spec`.

echasnovski commented:
@hrsh7th do you mind me asking what are your current plans for this?

hrsh7th commented:
Everyone uses LuaSnip, so it was a little less of a priority.
I think I'll raise the priority again and think about it.

clason commented:
> Everyone uses LuaSnip, so it was a little less of a priority.
> I think I'll raise the priority again and think about it.

Just to be clear: it is still a very high priority for Neovim to be able to expand LSP snippets (e.g., in omni-completion, or using `nvim-cmp`) without an external plugin. We consider this a requirement for claiming to fully support the LSP specification, which is an explicit goal.

(We also hope to be able to expose these methods to users wanting to write their own full-featured snippet plugin that can work with a personal snippet library and manual expansion, but this is secondary.)

textzenith commented:
This functionality would be very useful for me.

I'm writing a project in Dart / Flutter — a mobile app — which requires sticking together a lot of "widgets" that each have a million props to configure.

The omnifunc completes props so I don't have to remember exactly what attributes are owned by whichever widget. However, _built into_ the ordinary completion, for ordinary tokens, in the Dart LSP is snippet functionality.

A regular completion on "co" inside a Container object would produce "color: ," — adding a colon, a space, and appending a trailing comma. The idea is for the cursor to be placed before the comma so you can just insert whatever object is required by the class property (attribute), and it's all syntactically correct.

However, because the Omnifunc doesn't have any support for snippet-like behaviour, it ignores the added information from the language server, and just places the cursor after the comma. Meaning I have to either delete the comma and re-enter it later, or go into normal mode to backtrack.

Likewise, trying to complete certain class names auto-fills all their properties, but you have no way to tab between them and edit them, because no snippets. This is all in the base LSP completions and no snippet plugin is going to be able to override that.

Not complaining, just throwing in my 2 cents about why this would be extremely, extremely welcome for certain users!

How's this PR looking, shall I just merge it in my local copy and see how it goes?

hrsh7th committed:
feat(snippet): add basic implementation

hrsh7th committed:
feat(snippet): refactor
- more type annotation
- more comments

hrsh7th commented on test/functional/lua/snippet_spec.lua:
```diff
@@ -0,0 +1,259 @@
+  --   eq({
+  --     'console.warn();',
+  --   }, helpers.buf_lines(0))
+  -- end)
```
How do I write test for this?

I can't manage `vim.schedule` and `feedkeys`...

hrsh7th committed:
feat(snippet): support nested snippet expansion
- support `selection=exclusive`
- jump/merge are only allowed when the cursor is into snippet context
- normalize tabstop order as sequential number
- implement SnippetSession:merge

hrsh7th committed:
feat(snippet): improve selection=exclusive test

mortezadadgar commented:
any news on this pr? won't be part of 0.10 milestone?

hrsh7th commented:
This PR i no longer needed as #25301 has merged. Thank you for @MariaSolOs !

I'm sorry that my work was slow...

MariaSolOs commented:
@hrsh7th Oh my, please don't be sorry! I wouldn't have been able to implement the snippet API without this PR as a starting point, and you're still the main maintainer of one of the most crucial completion plugins in the entire ecosystem!
